File: code/csharp/DesignPatterns/flyweight/ConcreteBook.cs
namespace Flyweight
{
    public class ConcreteBook : Book
    {
        private readonly string _title;
        private readonly string _author;
        private bool _isAvailable = true;
        private int _id;

        public ConcreteBook(string title, string author, int id)
        {
            _title = title;
            _author = author;
            _id = id;
        }

        public void BorrowBook()
        {
            if (_isAvailable)
            {
                _isAvailable = false;
                System.Console.WriteLine($"Book borrowed: {_title} by {_author} book id: {_id}");
            }
            else
            {
                System.Console.WriteLine($"Book is currently not available: {_title}");
            }
        }

        public void ReturnBook()
        {
            _isAvailable = true;
            System.Console.WriteLine($"Book returned: {_title} by {_author} book id: {_id}");
        }

        public bool IsAvailable() => _isAvailable;
        public int GetId() => _id;
    }
}

File: code/python/flyweight/ConcreteBook.py
from Book import Book


class ConcreteBook(Book):
    def __init__(self, title, author, book_id):
        self.title = title
        self.author = author
        self._is_available = True  # Use a private attribute to manage the state
        self.id = book_id

    def borrow_book(self):
        if self.is_available():
            self._is_available = False
            print(f"Book borrowed: {self.title} by {self.author} book id: {self.id}")
        else:
            print(f"Book is currently not available: {self.title}")

    def return_book(self):
        self._is_available = True
        print(f"Book returned: {self.title} by {self.author} book id: {self.id}")

    def is_available(self):
        return self._is_available

    def get_id(self):
        return self.id
